### PR TITLE
chore(iast): refactor iast request context by core.context

### PR DESCRIPTION
**Pull request by avara1986**
This PR continues the work Christophe started in PR #10899, focusing on a refactor of the IAST context. Key highlights of this refactor include:

- Similar to PR #10899, an `IASTEnvironment` class has been introduced, which contains `request_enabled` and `IastSpanReporter`. These no longer depend on spans, only on the core context.
- The IAST context is still created in the Span Processor `on_span_start`.
- The context now finalizes with `with core.context_with_data`, and when the WSGI request ends, it calls the listener `context.ended.wsgi.__call__`. Previously, when calling Span Processor `on_span_finish`, the context was sometimes lost. Now, it ensures that the context hasn’t already been reported to avoid this issue, as seen in cases like the Django tests, where `Span Processor.on_span_finish` is called instead of `context.ended.wsgi.__call__`.
- The `thread_local struct ThreadContextCache_` has been temporarily removed to align C++ context behavior with Python’s. This is reflected in the test `test_oce_concurrent_requests_futures_in_spans`. That is a rollback of https://github.com/DataDog/dd-trace-py/pull/8452.
- The function `_patched_fastapi_function` and Header patching was removed.
- Additional validation points have been added for `is_iast_request_enabled` to avoid unnecessary operations.
- The logic from `AppSecIastSpanProcessor.is_span_analyzed` has been moved to `is_iast_request_enabled`.
- `_asm_request_context.listen` has been renamed to `_asm_request_context.asm_listen` to avoid confusion with other `listen` methods.
- `_asm_request_context.in_context` has been renamed to `_asm_request_context.in_asm_context` to differentiate it from `in_iast_context`.
- The import of `DDWaf_result` from `ddtrace.appsec._ddwaf` has been moved inside the `if TYPE_CHECKING` block to prevent circular imports caused by this refactor.
- IAST wrappers: check if wrapped function has a __func__ to skip `RecursionError: maximum recursion depth exceeded in comparison` in some Python versions https://gitlab.ddbuild.io/DataDog/apm-reliability/dd-trace-py/-/jobs/668530031
- re.Match is disabled due to segmentation faults and not valid ranges: task APPSEC-55281
Tests:
- GRPC tests are failing now task APPSEC-55239
- All IAST tests have been refactored to accommodate the new way of creating and using the context.
- The test `test_weak_hash_new_with_child_span` has been removed because we no longer depend on spans or child spans.
- The deprecated attribute `oce._enabled = True` has been removed.
- An IAST test in `tests/tracer/test_trace_utils.py` had to be updated and was moved to `taint_sinks/test_insecure_cookie.py`, so it requires validation from the @DataDog/apm-sdk-api-python team.
- Context creation/finalization has been updated in the benchmarks, requiring validation from the @DataDog/apm-core-python team.

## Checklist
- [x] PR author has checked that all the criteria below are met
- The PR description includes an overview of the change
- The PR description articulates the motivation for the change
- The change includes tests OR the PR description describes a testing strategy
- The PR description notes risks associated with the change, if any
- Newly-added code is easy to change
- The change follows the [library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html)
- The change includes or references documentation updates if necessary
- Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist
- [x] Reviewer has checked that all the criteria below are met 
- Title is accurate
- All changes are related to the pull request's stated goal
- Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- Testing strategy adequately addresses listed risks
- Newly-added code is easy to change
- Release note makes sense to a user of the library
- If necessary, author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

File: ddtrace/appsec/__init__.py [hunk 1]
```diff
@@ -3,9 +3,11 @@
...
     """Lazily load the appsec module listeners."""
-    from ddtrace.appsec._asm_request_context import listen
+    from ddtrace.appsec._asm_request_context import asm_listen
+    from ddtrace.appsec._iast._iast_request_context import iast_listen
 
     global _APPSEC_TO_BE_LOADED
     if _APPSEC_TO_BE_LOADED:
-        listen()
+        asm_listen()
+        iast_listen()
```
Comment (christophe-papazian) [inline ddtrace/appsec/__init__.py:12]: Shouldn't we keep the load_appsec function only for threats and put iast in its own `load_iast` function?

File: tests/appsec/iast/test_overhead_control_engine.py [hunk 2]
```diff
@@ -161,3 +169,33 @@ def test_oce_concurrent_requests(tracer, iast_span_defaults):
...
+    span_report = get_iast_reporter()
+
+    assert len(span_report.vulnerabilities)
+
+    assert len(results) == num_requests * 3
+    assert len(span_report.vulnerabilities) >= 1
+
+    # log_messages = [record.message for record in caplog.get_records("call")]
+    # if not any(message.startswith("[IAST] no vulnerability quota") for message in log_messages):
+    #     pytest.fail()
```
Comment (gnufede) [inline tests/appsec/iast/test_overhead_control_engine.py:201]:
```suggestion
```

File: ddtrace/appsec/_iast/_patch.py [hunk 3]
```diff
@@ -149,23 +136,22 @@ def _on_iast_fastapi_patch():
...
+    #     "starlette.datastructures",
+    #     "Headers.get",
+    #     functools.partial(if_iast_taint_returned_object_for, OriginType.HEADER),
+    # )
+    # try_wrap_function_wrapper(
+    #     "fastapi",
+    #     "Header",
+    #     functools.partial(_patched_fastapi_function, OriginType.HEADER),
+    # )
+    # _set_metric_iast_instrumented_source(OriginType.HEADER)
```
Comment (gnufede) [inline ddtrace/appsec/_iast/_patch.py:154]: Is this be related to the FastAPI issue? Maybe just remove it?
```suggestion
```

File: ddtrace/appsec/_iast/_taint_tracking/__init__.py [hunk 4]
```diff
@@ -139,6 +142,9 @@ def is_pyobject_tainted(pyobject: Any) -> bool:
 
 def taint_pyobject(pyobject: Any, source_name: Any, source_value: Any, source_origin=None) -> Any:
     # Pyobject must be Text with len > 1
```
Comment (gnufede) [inline ddtrace/appsec/_iast/_taint_tracking/__init__.py:144]: This comment is now far from the logic it explains

**Review by gnufede — APPROVED** [review]

**Review by mabdinur — APPROVED** [review]
Changes in `tests/tracer/test_trace_utils.py` LGTM. 

Also there are also a few more tests in `tests/tracer/` that are ASM specific (they are mostly ASM standalone tests). Would it make sense to move these tests to `tests/appsec/`? This could be part of a greater effort to decouple ASM standalone from core tracing logic but it's not a requirement (we should do this in a separate PR)?

**Comment by avara1986** [comment]
> Changes in `tests/tracer/test_trace_utils.py` LGTM.
> 
> Also there are also a few more tests in `tests/tracer/` that are ASM specific (they are mostly ASM standalone tests). Would it make sense to move these tests to `tests/appsec/`? This could be part of a greater effort to decouple ASM standalone from core tracing logic but it's not a requirement (we should do this in a separate PR)?

Great idea! I didn't see the standalone tests. My next PR is decouple ASM and IAST contexts and listen, I can decuple standalone tests too :) TY!!!

**Review by erikayasuda — APPROVED** [review]
Just reviewed `benchmarks/bm/utils.py`. LGTM!
